rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiHost.kt

```diff
@@ -28,85 +125,94 @@ class FsiHost(project: Project) : LifetimedProjectComponent(project) {
+      (lastFocusedSession?.let { if (it.isValid()) resolvedPromise(it) else null }
+        ?: tryCreateDefaultConsoleRunner()).onSuccess {
+        it.sendText(visibleText, fsiText)
+      }
     }
   }
 
-  fun resetFsiConsole(forceOptimizeForDebug: Boolean, attach: Boolean = false) {
-    synchronized(lockObj) {
-      if (consoleRunner?.isValid() == true) {
-        consoleRunner!!.processHandler.destroyProcess()
+  fun resetFsiDefaultConsole() {
+    synchronized(lockObject) {
+      val session = sessions.getOrDefault("", null)
+      if (session != null) {
+        if (session.isValid()) session.stop()
+        sessions.remove("")
       }
-      tryCreateConsoleRunner(forceOptimizeForDebug, attach)
+      tryCreateDefaultConsoleRunner()
     }
   }
 
-  private fun getOrCreateConsoleRunner(forceOptimizeForDebug: Boolean): Promise<FsiConsoleRunner> {
-    if (consoleRunner?.isValid() == true) {
-      val result = AsyncPromise<FsiConsoleRunner>()
-      result.setResult(consoleRunner!!)
-      return result
+  private val sessions = mutableMapOf<String?, FsiConsoleRunnerBase>()
+  private val lockObject = Object()
+  var lastFocusedSession: FsiConsoleRunnerBase? = null
+
+  private fun getOrCreateConsoleRunner(
+    sessionId: String,
+    factory: () -> FsiConsoleRunnerBase
+  ): FsiConsoleRunnerBase {
+    val session = sessions.getOrPut(sessionId) {
+      val session = factory()
+      session.initAndRun()
+      session.processHandler.addProcessListener(object : ProcessAdapter() {
+        override fun processTerminated(event: ProcessEvent) {
+          synchronized(lockObject) {
+            sessions.remove(sessionId)
+          }
+        }
+      })
+      session.getRunContentDescriptor().preferredFocusComputable.compute().addFocusListener(object : FocusListener {
+        override fun focusGained(e: FocusEvent?) {
+          lastFocusedSession = session
+        }
+
+        override fun focusLost(e: FocusEvent?) {}
+      })
+      session
     }
-
-    return tryCreateConsoleRunner(forceOptimizeForDebug)
+    lastFocusedSession = session
+    return session
   }
 
-  private fun createConsoleRunner(
-    sessionInfo: RdFsiSessionInfo,
-    forceOptimizeForDebug: Boolean,
-    attach: Boolean
-  ): FsiConsoleRunner? =
-    synchronized(lockObj) {
-      // Might have already been created.
-      if (consoleRunner?.isValid() == true)
-        return consoleRunner
-
-      val fsiPath = sessionInfo.fsiPath
-      if (sessionInfo.runtime != RdFsiRuntime.Core && !FileUtil.exists(fsiPath)) {
-        notifyFsiNotFound(fsiPath)
-        return null
-      }
-
-      val runner = FsiConsoleRunner(sessionInfo, this, forceOptimizeForDebug)
-      runner.initAndRun()
-      this.consoleRunner = runner
-
-      if (attach)
-        runner.attachToProcess()
-
-      return runner
+  private fun tryCreateDefaultConsoleRunner(): Promise<FsiConsoleRunnerBase> {
+    val result = AsyncPromise<FsiConsoleRunnerBase>()
```

Comment:
`AsyncPromise` is deprecated, so definitely let's migrate to coroutines.

Or… why is it `AsyncPromise` here exactly? To me, it looks like this function only ever resolves it on the synchronous path. So, maybe get rid of the promise altogether?